Vector implementation

garfr opened the pull request:
# Description

Provides a basic and minimal implementation of a generic resizable array.  It provides these basic functions:

* init_vector
* init_vector_of_size
* push_vector
* index_vector
* free_vector

The implementation generally opts to return NULL on errors, but I also considered returning more useful error codes.  However, I think the best way to do this is with a global error queue, which requires other functionality outside this issue.  This can easily be added in backwards, if the project decides to go that direction.

Closes #12

gavinratcliff committed:
Added a CMakeLists.txt to the utils directory (#6, #7, #9, #10, #11, #12)

gavinratcliff committed:
Designed interface for vector (#12)

gavinratcliff committed:
Implemented interface for vector (#12)

gavinratcliff committed:
Wrote basic unit tests for vector (#12)

gavinratcliff committed:
Updated .gitignore to avoid temp files

gavinratcliff committed:
clang-formated vector source files (#12)

gavinratcliff committed:
Added doxygen documentation (#12)

gavinratcliff committed:
Renamed vector field, refactored documentation (#12)

gavinratcliff committed:
Added testing to cmake (#12)

gavinratcliff committed:
Revamped testing to use new methodology (#12)

gavinratcliff committed:
Fixed several comments, improved docs (#12)

gavinratcliff committed:
Synchronized docs between vector header and source (#12)

bama4 commented on utils/include/vector.h:
```diff
@@ -0,0 +1,65 @@
+/*!
+ * @brief     Initializes a new vector of size 0
+ *
+ * @param     item_size Size of the items to be stored in the vector in bytes,
```
The direction needs to be specified on all params, right?

garfr replied:
I'm not sure... I know its legal to not specify.  I think its only necessary when the param is a pointer, as passing a ``size_t`` will never be an out param.  Its also more important in Javadoc, which is where the style comes from, because all non built-in types are references.

bama4 commented on utils/src/vector.c:
```diff
@@ -0,0 +1,152 @@
+
+        /* Check for memory allocation failure */
+        if (vec->contents == NULL) {
+            return NULL;
```
You need to free vec, if this turns out to be true, else there will be a memory leak

garfr replied:
Good catch!

bama4 commented on utils/test/test_vector.c:
```diff
@@ -0,0 +1,116 @@
+    printf("Running %s\n", __func__);
+    int result = SUCCESS;
+
+    vector_t* test_vec = init_vector_of_size(1, 0);
```
Need a comment for each unit test describing what it is testing

bama4 commented on utils/test/test_vector.c:
```diff
@@ -0,0 +1,116 @@
+    printf("Running %s\n", __func__);
+    int result = SUCCESS;
+
+    result += test_init_vector_of_size();
```
Should be `result &=` so that the value of result stays either `0` or `-1`

garfr replied:
Oops, mispelling.

bama4 commented:
I suggest you run `valgrind` to make sure there are no memory leaks if you have not already

bama4 commented on utils/include/vector.h:
```diff
@@ -0,0 +1,65 @@
+    size_t capacity;
+    size_t used;
+    size_t item_size;
+    char* contents;
```
Based on the way you are accessing the vector through accessors such as `index_vector`, it appears that `contents` is meant to be a private member? If so maybe indicate this by using `_contents`

garfr replied:
Contents should be accessable fully as well.  index_vector just bounds checks and takes into account the item_size.

bama4 replied:
Ok. We should just keep in mind that we want to be consistent in how we are accessing the vector, if there are accessors available and do checking for us, we may want to use those outside of this library

garfr replied:
Sounds good. I will add an ``index_bytes`` and ``get_contents`` function and rename ``contents`` to ``_contents``

bama4 commented on utils/src/vector.c:
```diff
@@ -0,0 +1,152 @@
+ *
+ * @param[in] vec Vector to be destroyed
+ */
+void free_vector(vector_t* vec) {
```
Looks like the way you are using a vector in the program, you are not doing pass-by-reference here.
The parameter should be `vector_t ** vec` to indicate that you are passing by reference.

garfr replied:
This seems like pass by reference?  Also, passing a with 2 levels of indirection seems silly, as the only benefit is that you can set the 2nd pointer to NULL.  Its also extremely minimal, but 2 memory accesses is slower than 1,

bama4 replied:
So in your tests, you initialize for example, a vector * test_vector, and then you call free_vector(test_vector) but because the parameter is vector * instead of vector **, the pointer is being passed by value, so you would not be able to set the freed pointer to NULL unless you did a pass by reference: https://stackoverflow.com/questions/6054271/freeing-pointers-from-inside-other-functions-in-c

bama4 replied:
So we would want to set the pointer to NULL after free

garfr replied:
Hmm.  Passing with a double pointer seems obtuse.  Most libraries that have functions that free objects use just one pointer and leave the programmer to decide if they want to set the pointer to NULL.

bama4 replied:
If passing a double pointer seems icky, then maybe we want to not use `vector_t *` and instead initialize vectors with `vector_t`. This way we can perform a pass by reference when we want to free the vector contents...what do you think?

gavinratcliff committed:
Added get_contents, index_bytes changed contents to _contents (#12)

mend

Added get_contents, index_byte and changed contents to _contents (#12)

Clang-formatted source

gavinratcliff committed:
Fixed possible memory leak in init_vector_of_size (#12)

gavinratcliff committed:
Added comments to unit tests (#12)

bama4 commented on utils/test/test_vector.c:
```diff
@@ -0,0 +1,133 @@
+    // Test indexing first item
+    int* ret = (int*)index_vector(vec, 0);
+    ASSERT_NOT_NULL(ret);
+    ASSERT_EQUAL(*ret, 1);
```
You never test that used or size members of vector_t are updated

bama4 commented on utils/src/vector.c:
```diff
@@ -0,0 +1,183 @@
+ * @return        Returns 1 on success or 0 on failure
+ */
+static int extend_vector(vector_t* vec, size_t change) {
+    if (vec->used + change <= vec->capacity) {
```
You never check to see if the vector_t parameter is NULL in any of the functions. I think a NULL check would be good

bama4 commented on utils/src/vector.c:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @param[in]  item Pointer to the item to be deep copied into the vector
+ *
+ * @return     1 on success, or 0 on failure
```
Just realized...the convention is 0 on success and -1 on failure for many C functions returning an int. For functions returning a pointer it is a valid pointer on success and NULL on failure. We should adhere to convention

bama4 commented on utils/include/vector.h:
```diff
@@ -0,0 +1,88 @@
+#include <stddef.h>
+
+typedef struct {
+    size_t capacity;
```
I like this implementation of vector, very easy to read the implementation and internals

gavinratcliff committed:
Made index_byte return pointer to byte (#12)

gavinratcliff committed:
Updated interface to return -1 on fail and 0 on success (#12)

gavinratcliff committed:
Added missing tests in unit tests (#12)

gavinratcliff committed:
Added null checks (#12)

gavinratcliff committed:
Fixed memory leak in test_index_vector (#12)

bama4 commented:
Output of valgrind:
```
~/REPOS/simple-http-server/utils/build$ valgrind --leak-check=full ./bin/utils_test
==13579== Memcheck, a memory error detector
==13579== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==13579== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==13579== Command: ./bin/utils_test
==13579== 
Running tests for utils library
Running test_vector
Running test_init_vector_of_size
Running test_init_vector
Running test_push_vector
Running test_index_vector
Vector test SUCCESS
==13579== 
==13579== HEAP SUMMARY:
==13579==     in use at exit: 32 bytes in 1 blocks
==13579==   total heap usage: 11 allocs, 10 frees, 1,739 bytes allocated
==13579== 
==13579== 32 bytes in 1 blocks are definitely lost in loss record 1 of 1
==13579==    at 0x4C31B0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==13579==    by 0x1088A1: init_vector_of_size (in /home/ubuntu/REPOS/simple-http-server/utils/build/bin/utils_test)
==13579==    by 0x10895B: init_vector (in /home/ubuntu/REPOS/simple-http-server/utils/build/bin/utils_test)
==13579==    by 0x1091D3: test_index_vector (in /home/ubuntu/REPOS/simple-http-server/utils/build/bin/utils_test)
==13579==    by 0x10947E: test_vector (in /home/ubuntu/REPOS/simple-http-server/utils/build/bin/utils_test)
==13579==    by 0x108BA8: main (in /home/ubuntu/REPOS/simple-http-server/utils/build/bin/utils_test)
==13579== 
==13579== LEAK SUMMARY:
==13579==    definitely lost: 32 bytes in 1 blocks
==13579==    indirectly lost: 0 bytes in 0 blocks
==13579==      possibly lost: 0 bytes in 0 blocks
==13579==    still reachable: 0 bytes in 0 blocks
==13579==         suppressed: 0 bytes in 0 blocks
==13579== 
==13579== For counts of detected and suppressed errors, rerun with: -v
==13579== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
